File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.*;
import java.util.*;
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        long M = Long.parseLong(st.nextToken());

        long[] A = new long[N];
        st = new StringTokenizer(br.readLine());
        long hi = 0;
        for(int i=0;i<N;i++) {
            A[i] = Long.parseLong(st.nextToken());
            if(A[i] > hi) hi = A[i];
        }

        long lo = 0, ans=0;
        while(lo <= hi) {
            long mid = (lo + hi) >>> 1; // (lo + hi) / 2
            long got = 0;

            for(int i=0;i<N;i++) {
                if(A[i] > mid) got += (A[i] - mid);
            }
            
            if(got >= M) {
                ans = mid;
                lo = mid+1;
            } else {
                hi = mid - 1;
            }
        }

        System.out.println(ans);
    }
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    long long M;
    cin >> N >> M;

    vector<long long> A(N);
    long long hi = 0;
    for (int i = 0; i < N; i++)
    {
        cin >> A[i];
        if (A[i] > hi)
            hi = A[i];
    }

    long long lo = 0, ans = 0;
    while (lo <= hi)
    {
        long long mid = (lo + hi) >> 1; // (lo+hi) / 2
        long long got = 0;
        for (int i = 0; i < N; i++)
        {
            if (A[i] > mid)
                got += (A[i] - mid);
        }

        if (got >= M)
        {
            ans = mid;
            lo = mid + 1;
        }
        else
        {
            hi = mid - 1;
        }
    }

    cout << ans;
}